fork-off/src/main.rs

```diff
@@ -12,168 +19,336 @@ pub struct Config {
-                .expect("Not a string")
-                .to_owned(),
-        );
+    pub async fn get_full_state(&self, num_workers: u32) -> Storage {
+        let best_block = self.get_most_recent_block().await;
+        let keys = self.get_all_keys(best_block.clone()).await;
+        self.get_values(keys, best_block, num_workers).await
     }
-    None
 }
 
-async fn get_value(http_rpc_endpoint: &str, key: &str) -> String {
-    let response: Value = reqwest::Client::new()
-        .post(http_rpc_endpoint)
-        .json(&serde_json::json!({
-            "jsonrpc": "2.0",
-            "id": 2,
-            "method": "state_getStorage",
-            "params": [ &key ]
-        }))
-        .send()
-        .await
-        .expect("Storage request has failed")
-        .json()
-        .await
-        .expect("Could not deserialize response as JSON");
+fn save_snapshot_to_file(snapshot: Storage, path: String) {
+    let data = serde_json::to_vec_pretty(&snapshot).unwrap();
+    info!(
+        "Writing snapshot of {} key-val pairs and {} total bytes",
+        snapshot.len(),
+        data.len()
+    );
+    write_to_file(path, &data);
+}
 
-    debug!(target: "fork", "get_value response: {}", response);
+fn read_snapshot_from_file(path: String) -> Storage {
+    let snapshot: Storage =
+        serde_json::from_str(&fs::read_to_string(&path).expect("Could not read snapshot file"))
+            .expect("could not parse from snapshot");
+    info!("Read snapshot of {} key-val pairs", snapshot.len());
+    snapshot
+}
 
-    response["result"]
-        .as_str()
-        .expect("Not a string")
-        .to_owned()
+fn is_prefix_of(shorter: &str, longer: &str) -> bool {
+    longer.starts_with(shorter)
 }
 
-async fn get_chain_state(
-    http_rpc_endpoint: &str,
-    hashed_prefixes: &[String],
-) -> Vec<(String, String)> {
-    stream::iter(hashed_prefixes.to_owned())
-        .then(|prefix| async move {
-            // collect storage pairs for this prefix
-            let mut pairs = vec![];
-            let mut first_key = get_key(http_rpc_endpoint, &prefix, None).await;
-            debug!(target: "fork", "hashed prefix: {}, first key: {:?}", &prefix, &first_key);
-
-            while let Some(key) = first_key {
-                let value = get_value(http_rpc_endpoint, &key).await;
-                pairs.push((key.clone(), value));
-                first_key = get_key(http_rpc_endpoint, &prefix, Some(&key)).await;
-                debug!(target: "fork", "hashed prefix: {}, next key: {:?}", &prefix, &first_key);
+fn combine_states(mut state: Storage, initial_state: Storage, pallets: Vec<String>) -> Storage {
+    let pallets_prefixes: Vec<(String, String)> = pallets
+        .iter()
+        .map(|pallet| {
+            let hash = format!("0x{}", prefix_as_hex(pallet));
+            (pallet.clone(), hash)
+        })
+        .collect();
+    let mut cnt_removed_per_pallet: HashMap<String, usize> = pallets_prefixes
```

Comment:
can we just call it with full name `removed_per_pallet_count` / `added_per_pallet_count` ?

Reply:
For you, done :)